### PR TITLE
feat(eowc): enable EOWC mode HashAgg

### PR DESCRIPTION
**Pull request by richardchien**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

As title.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- ~~[ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).~~
- ~~[ ] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)~~
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/frontend/src/optimizer/plan_node/stream_hash_agg.rs [hunk 1]
```diff
@@ -124,11 +139,9 @@ impl PlanTreeNodeUnary for StreamHashAgg {
...
 
     fn clone_with_input(&self, input: PlanRef) -> Self {
-        let logical = generic::Agg {
-            input,
-            ..self.logical.clone()
-        };
-        Self::new(logical, self.vnode_col_idx, self.row_count_idx)
+        let mut new = self.clone();
+        new.logical.input = input;
+        new
```
Comment (st1page) [inline src/frontend/src/optimizer/plan_node/stream_hash_agg.rs:144]: we just want to maintain the eowc property here, but ti will skip all property's derivation including order and distribution. I think they are dangerous, maybe we should not do that.

File: src/frontend/src/optimizer/plan_node/logical_agg.rs [hunk 2]
```diff
@@ -1099,22 +1101,37 @@ impl ToStream for LogicalAgg {
...
-        } else if let Some(final_agg) = stream_agg.as_stream_hash_agg() {
-            final_agg.agg_calls()
+        let (plan, n_final_agg_calls) = if let Some(final_agg) = plan.as_stream_simple_agg() {
+            if eowc {
+                return Err(ErrorCode::InvalidInputSyntax(
+                    "`EMIT ON WINDOW CLOSE` cannot be used for aggregation without `GROUP BY`"
+                        .to_string(),
+                )
+                .into());
+            }
```
Comment (st1page) [inline src/frontend/src/optimizer/plan_node/logical_agg.rs:1113]: It is ok for now before we derive watermark for the append only `max` aggCall

**Review by st1page — COMMENTED** [review]
please add some planner test

File: src/stream/src/executor/sort_buffer.rs [hunk 3]
```diff
@@ -227,3 +231,32 @@ impl<S: StateStore> SortBuffer<S> {
         Ok(())
     }
 }
+
+fn fix_row<S: StateStore>(
```
Comment (chenzl25) [inline src/stream/src/executor/sort_buffer.rs:235]: Could you please shed some light on it?

Comment (richardchien) [reply src/stream/src/executor/sort_buffer.rs]: Renamed the function and added some comment.

File: src/stream/src/executor/hash_agg.rs [hunk 4]
```diff
@@ -199,6 +202,11 @@ impl<K: HashKey, S: StateStore> HashAggExecutor<K, S> {
             &args.agg_calls,
             Some(&args.extra.group_key_indices),
         );
+        // NOTE: we assume the prefix of table pk is exactly the group key
+        let group_key_table_pk_projection = args.result_table.pk_indices()
+            [..args.extra.group_key_indices.len()]
+            .to_vec()
+            .into();
```
Comment (chenzl25) [inline src/stream/src/executor/hash_agg.rs:209]: Does it affect correctness? I mean if the watermark column is the second column of the group by keys.

Comment (richardchien) [reply src/stream/src/executor/hash_agg.rs:209]: The projection is actually to handle the cases where watermark column is the second inside group key cols. See the following:

https://github.com/risingwavelabs/risingwave/blob/7defa87ff38bc715b8ebbb26408492a3a7ad5498/src/frontend/src/optimizer/plan_node/generic/agg.rs#L278-L329

When grouped by a window watermark column, the column will be added to the table schema in `group_key` bitset order but in another order in table pk. Executor assumes that the window column is the first col in table pk but doesn't assume the position of it in table schema.

**Pull request merged**